src/relay/op/type_relations.cc

```diff
@@ -0,0 +1,135 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file type_relations.cc
+ * \brief A set of utilities and common functionality
+ * for type relations.
+ */
+#include <tvm/relay/expr.h>
+#include <tvm/relay/logging.h>
+#include <tvm/relay/op.h>
+#include "../pass/incomplete_type.h"
+
+namespace tvm {
+namespace relay {
+
+TensorType as_ttype(const Type& t) {
+  if (auto tt_node = t.as<TensorTypeNode>()) {
+    return GetRef<TensorType>(tt_node);
+  } else {
+    return TensorType(nullptr);
+  }
+}
+
+// TODO(@jroesch) what size value do we extract?
+int to_int(const tvm::Expr& e) {
+  auto imm = e.as<tvm::ir::IntImm>();
+  CHECK(imm);
+  std::cout << "TYPE: " << imm << imm->type << std::endl;
+  return imm->value;
+}
+
+Array<Type> IdentityRel(const Array<Type>& types, int num_args) {
+  CHECK(types.size() == 2);
+  auto t1 = as_ttype(types[0]);
+  if (t1 && types[1].as<IncompleteTypeNode>()) {
+    return {t1, t1};
+  } else {
+    return types;
+  }
+}
+
+static Type ConcreteBroadcast(const TensorType& t1, const TensorType& t2,
+                              DataType output_dtype) {
+  RELAY_LOG(INFO) << "ConcreteBroadcast: t1=" << t1 << " t2=" << t2
+                  << std::endl;
+  auto sh1 = t1->shape;
+  auto sh2 = t2->shape;
+  RELAY_LOG(INFO) << "ConcreteBroadcast: sh1=" << sh1 << " sh2=" << sh2
+                  << std::endl;
+  if (sh1.size() == 0 && sh2.size() == 0) {
+    return TensorTypeNode::make({}, output_dtype);
+    // We have non-zero shapes so broadcast rules apply.
+  } else {
+    auto suffix_len = static_cast<int>(std::min(sh1.size(), sh2.size()));
+    auto full_len = static_cast<int>(std::max(sh1.size(), sh2.size()));
+
+    std::cout << "Length" << suffix_len << full_len
+              << (full_len - suffix_len - 1) << std::endl;
+    auto lower_bound = full_len - suffix_len - 1;
+
+    for (int64_t i = full_len - 1; i > lower_bound; i--) {
```

Comment:
this will out of range immediately on lb of different size

Reply:
suppose sh1.size() is 10, sh2.size() is 5
lb = 4
i originally = 9
will out of bound